### PR TITLE
apiv2,server: filter expensive db queries from sql-over-http

### PR DESCRIPTION
**Pull request by dhartunian**
Retrieving replica counts and region data via the `SHOW RANGES FROM DATABASE` and `SHOW RANGES FROM TABLE` queries is extremely costly, especially on a large cluster.

Executing these queries should not be done automatically and it's unclear if the cost of computation is worth the data that the customer can see. It would be better for a customer to explicitly opt-in to computing this data in special cases.

A cluster setting (`ui.database_locality_metadata.enabled`) has been introduced to allow customers to turn this functionality off if they're managing a large fleet where it's difficult to prevent users from using this feature. By default this data is enabled.

The implementation is admittedly a hack, designed to reduce the size of this diff and the risk of the backport. Implementing this change purely on the client-side resulted in too much code needing to be changed as we would both need to conditionally change the queries being used to load data, and gate them on cluster setting being loaded successfully prior to the SQL requests being made.

Resolves: CRDB-42482

Release note (ops change, ui change): A new cluster setting `ui.database_locality_metadata.enabled` allows operators to disable loading extended database and table region information in the DB Console Database and Table pages. This information can cause significant CPU load on large clusters with many ranges. Versions of this page from 24.3 onwards do not have this problem. If customers will require this data, they can use the `SHOW RANGES FROM {DATABASE| TABLE}` query via SQL to compute on-demand.

### CODE REVIEW

File: pkg/server/api_v2_sql_test.go [hunk 1]
```diff
@@ -130,3 +130,36 @@ func getErrorResponse(t *testing.T, r []byte) (code, message string) {
 	require.NoError(t, err)
 	return er.Error.Code, er.Error.Message
 }
+
+func TestLocalityMetadataEnabledFilter(t *testing.T) {
```
Comment (xinhaoz) [inline pkg/server/api_v2_sql_test.go:134]: Can I request some additional tests executing these queries with the cluster setting off? Just to see that the result is what we expect. Or a loom showing what the console page looks like with the cluster setting on.

**Review by angles-n-daemons — CHANGES_REQUESTED** [review]
Just so I understand, where are the automatic `SHOW RANGES` queries coming from? What makes removing them a challenge?

Do you have a list of (non-test) locations where it is used and what its response is used for?

Couple additional questions in the PR, looks simple which is good, but want to explore impact a little more.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133075#-:-O9uT0nwEDDeFA52Y4G4:b-4cqp3n)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @dhartunian, @kyle-a-wong, and @xinhaoz)

___
*[`pkg/server/api_v2_sql.go` line 343 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133075#-O9uSuW4Eb8kl_t9hyTk:-O9uSuW4Eb8kl_t9hyTl:bjjime0) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0fc059fc5b46425554f66d20ad18400d7a3f32f/pkg/server/api_v2_sql.go#L343)):*
> ```Go
> 				// continue and return a matchin list of responses. Client-side
> 				// code is resilient to empty results.
> 				s.SQL = "SELECT 1"
> ```

The shape of the response here won't match the shape of the query - is this okay?
___
*[`pkg/server/api_v2_sql.go` line 630 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133075#-O9uS0rzA2DC0fMUs8fv:-O9uS0rzA2DC0fMUs8fw:b-i8evwn) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0fc059fc5b46425554f66d20ad18400d7a3f32f/pkg/server/api_v2_sql.go#L630)):*
> ```Go
> // execute and can be disabled by the operator. See the Note above for
> // an explanation of this hack.
> var localityMetadataQueryRegexp = regexp.MustCompile(`SELECT.+AS store_ids\s+FROM \[SHOW RANGES FROM (DATABASE|TABLE) .+\], unnest\(replicas\) AS unnested`)
> ```

Is this the minimum regexp which will match? It seems like if `SHOW RANGES FROM (DATABASE|TABLE)` is the offending operation, the expression could be just that, flanked by `.+` on either side if necessary

File: pkg/server/api_v2_sql.go [hunk 2]
```diff
@@ -597,3 +621,10 @@ func (j jsonError) MarshalJSON() ([]byte, error) {
 	pqErr := pgerror.Flatten(j.error)
 	return gojson.MarshalIndent(pqErr, "", "")
 }
+
+// This regexp is constructed from client-side queries written in
+// `databaseDetailsApi.ts` and `tableDetailsApi.ts`. It's meant to
+// filter out queries that use `SHOW RANGES` which are expensive to
+// execute and can be disabled by the operator. See the Note above for
+// an explanation of this hack.
+var localityMetadataQueryRegexp = regexp.MustCompile(`SELECT.+AS store_ids\s+FROM \[SHOW RANGES FROM (DATABASE|TABLE) .+\], unnest\(replicas\) AS unnested`)
```
Comment (xinhaoz) [inline pkg/server/api_v2_sql.go:630]: Just wondering, do we need to be this specific for the regexp? Why not just match on SHOW RANGES and an application name? I don't really mind either approach though.

**Review by dhartunian — COMMENTED** [review]
> Just so I understand, where are the automatic `SHOW RANGES` queries coming from? What makes removing them a challenge?
> 
> Do you have a list of (non-test) locations where it is used and what its response is used for?

@angles-n-daemons I linked to the two client-side typescript files that contain the queries I'm trying to filter out as well as a recording of the resulting behavior in a response to your question below.

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133075#-:-OAcD1A64-5qAGXC-CWp:bgth91s)* status: <span>:shipit:</span> complete! 0 of 0 LGTMs obtained (waiting on @angles-n-daemons, @kyle-a-wong, and @xinhaoz)

___
*[`pkg/server/api_v2_sql.go` line 343 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133075#-O9uSuW4Eb8kl_t9hyTk:-O9ytOyY3gghlerpFv94:biglakh) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0fc059fc5b46425554f66d20ad18400d7a3f32f/pkg/server/api_v2_sql.go#L343)):*
<details><summary><i>Previously, angles-n-daemons (Brian Dillmann) wrote…</i></summary><blockquote>

The shape of the response here won't match the shape of the query - is this okay?
</blockquote></details>

The response is parsed by the frontend dynamically based on returned columns. That should be fine. Here's a loom with the behavior so you can verify:

https://www.loom.com/share/9a25c888a262478ab4092d719075a49a

I can't remove the offending query because client-side code expects responses to contain same number of results as queries that were dispatched.
___
*[`pkg/server/api_v2_sql.go` line 630 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133075#-O9uS0rzA2DC0fMUs8fv:-O9ytUa0AV_MeD-6srBh:b-ce6ec5) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0fc059fc5b46425554f66d20ad18400d7a3f32f/pkg/server/api_v2_sql.go#L630)):*
<details><summary><i>Previously, angles-n-daemons (Brian Dillmann) wrote…</i></summary><blockquote>

Is this the minimum regexp which will match? It seems like if `SHOW RANGES FROM (DATABASE|TABLE)` is the offending operation, the expression could be just that, flanked by `.+` on either side if necessary
</blockquote></details>

I wrote this to match the exact queries that are written in the frontend in these files:

https://github.com/cockroachdb/cockroach/blob/master/pkg/ui/workspaces/cluster-ui/src/api/databaseDetailsApi.ts#L379-L380

https://github.com/cockroachdb/cockroach/blob/master/pkg/ui/workspaces/cluster-ui/src/api/tableDetailsApi.ts#L449-L450

My intent was not to match arbitrary clients querying this API, although based on @xinhaoz's comment, perhaps a simpler regexp is better in this case.
___
*[`pkg/server/api_v2_sql.go` line 630 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133075#gh-1813270669:-OAcCEy-B5keHAD3Krvn:b-o60fge) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0fc059fc5b46425554f66d20ad18400d7a3f32f/pkg/server/api_v2_sql.go#L630)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Just wondering, do we need to be this specific for the regexp? Why not just match on SHOW RANGES and an application name? I don't really mind either approach though.
</blockquote></details>

Since both you and Brian have brought it up, I've simplified the regexp. Given that this change is affecting older versions only, it's very unlikely that new usage will end up bumping into this limitation.
___
*[`pkg/server/api_v2_sql_test.go` line 134 at r1](https://reviewable.io/reviews/cockroachdb/cockroach/133075#gh-1813262816:-O9yu9byBTxGZYmfy-ST:bj5gbu2) ([raw file](https://github.com/cockroachdb/cockroach/blob/f0fc059fc5b46425554f66d20ad18400d7a3f32f/pkg/server/api_v2_sql_test.go#L134)):*
<details><summary><i>Previously, xinhaoz (Xin Hao Zhang) wrote…</i></summary><blockquote>

Can I request some additional tests executing these queries with the cluster setting off? Just to see that the result is what we expect. Or a loom showing what the console page looks like with the cluster setting on.
</blockquote></details>

Done.  Tests added to the datadriven suite for sql endpoint.

Loom: https://www.loom.com/share/9a25c888a262478ab4092d719075a49a

**Review by angles-n-daemons — APPROVED** [review]
:lgtm: can the follow up to remove this be a fast follow? since it sounds like it isn't necessary in 24.3 onward?

> *[Reviewable](https://reviewable.io/reviews/cockroachdb/cockroach/133075#-:-OAcrSSn5kRvwZjoXSSo:b-iiq5lg)* status: <span>:shipit:</span> complete! 1 of 0 LGTMs obtained (waiting on @kyle-a-wong and @xinhaoz)

**Comment by dhartunian** [comment]
ack @angles-n-daemons, I will submit PR to `master` after backports are open to remove this.

bors r+

**Pull request merged**